src/interface/swarm.hpp

```diff
@@ -0,0 +1,150 @@
+//========================================================================================
+// (C) (or copyright) 2020. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_HPP_
+#define INTERFACE_SWARM_HPP_
+
+///
+/// A swarm contains all particles of a particular species
+/// Date: August 21, 2019
+
+#include "basic_types.hpp"
+#include "bvals/cc/bvals_cc.hpp"
+#include "metadata.hpp"
+#include "parthenon_arrays.hpp"
+#include "variable.hpp"
+#include <array>
+#include <cstdint>
+#include <list>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace parthenon {
+class MeshBlock;
+
+enum class PARTICLE_STATUS { UNALLOCATED, ALIVE, DEAD };
+
+class Swarm {
+ public:
+  MeshBlock *pmy_block = nullptr; // ptr to MeshBlock
+
+  Swarm(const std::string label, const Metadata &metadata, const int nmax_pool_in = 3);
+
+  // TODO BRR This should really be const... mask_ is managed internally
+  /// Get mask array for active particles
+  ParticleVariable<bool> &GetMask() { return mask_; }
+
+  /// Get the mask array of active particles that are marked for removal
+  ParticleVariable<bool> &GetMarkedForRemoval() { return marked_for_removal_; }
+
```

Comment:
On first look this rubs me the wrong way. Typically you don't want to return a non const reference from a class because it breaks encapsulation. I will need to look closer to see if your use case is ok but my preference would be that these would be:
```
const ParticleVariable<bool> & GetMask() const noexcept { return mask_; }
```

The same for the GetMarked method. You could then have a separate setter if you need to asctually change the mask_. 

The other option is to simply return a copy instead of a reference.

Reply:
So I strongly agree with you that these two should be `const` methods. `mask_` and `marked_for_removal_` are both managed internally and users should not be manually editing these variables. However, it looks like there are issues with just returning these as const. For example, `GetHostMirrorAndCopy` is not a `const` method so I can't do `GetMask().GetHostMirrorAndCopy()` which I need to do to use these variables at all. I think this should be addressed, but I'm not immediately sure how to. If I work through it in the next day or two I'll update this PR

Reply:
If you are having to touch code outside the scope of this PR I would not worry about it, just create an issue and link here, otherwise this will never get merged. If it is a simple change then ya go ahead. Most of the issues I raised are simply bringing light to issues that should be addressed but it doesn't mean they need to be addressed by you if you did not introduce them. Thanks for looking into this.

Reply:
#315 created, thanks